### ISSUE TITLE
Use cases that are not well served by service workers

### ISSUE THREAD

**Issue by dotproto** [w3c/webextensions#72]
Near the end of the [2021-08-19](https://github.com/w3c/webextensions/blob/main/_minutes/2021-08-19-wecg.md#:~:text=features%20that%20are%20in%20scope%20in%20this%20cg) meeting we briefly discussed (but didn't capture) the idea of collecting background page use cases and design patterns that are not well served by service workers. My hope is that by collecting these use cases in a well known location, community members and browser vendors we will be able to more concretely discuss the challenges posed by this new background context and explore potential solutions. 

Please use this issue to link to use use cases that are impossible to accomplish with or not well supported by service workers. If you'd like to report a new use case, please [create a new issue](https://github.com/w3c/webextensions/issues/new) and reference this one.

## Use cases

Specific ways that background pages are used that cannot be accomplished with a service worker.

* HTML/XML parsing
  * [Issue 1056354: Add support for parsing XML and HTML documents to Manifest V3](https://bugs.chromium.org/p/chromium/issues/detail?id=1056354)
  * [Issue 1051597: Extension Manifest V3 support for SingleFileZ and similar extensions](https://bugs.chromium.org/p/chromium/issues/detail?id=1051597)
* Background page interaction
  * [Issue 1128240: Background page loading & manipulation in service worker browser extensions](https://bugs.chromium.org/p/chromium/issues/detail?id=1128240)
* Audio playback
  * [Issue 1131236: Playing audio in service worker-based browser extensions](https://bugs.chromium.org/p/chromium/issues/detail?id=1131236)
* Clipboard access
  * [Issue 1160302: Cannot read clipboard from service worker in a MV3 chrome extension](https://bugs.chromium.org/p/chromium/issues/detail?id=1160302)
* Blob URLs
  * Issue 1051597, [comment #5](https://bugs.chromium.org/p/chromium/issues/detail?id=1051597#c5)+ 
* Requesting file:-scheme resources (aka file:// URLs)
  * Issue 1051597, [comment #12](https://bugs.chromium.org/p/chromium/issues/detail?id=1051597#c12)+ – Discussion of file resource request limitations in Fetch
* WebRTC streams
  * Extension "Snowflake" https://github.com/w3c/webrtc-extensions/issues/77#issuecomment-912667886
  * [Issue 1207214: RTCDataChannel for ManifestV3](https://bugs.chromium.org/p/chromium/issues/detail?id=1207214)
* WebSockets
  * chromium-extensions: [MV3 Service Workers Don't Support Websocket](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/o6xecpul22M/m/Pv07mPeaAwAJ)
  * https://github.com/w3c/webrtc-extensions/issues/77
* Native messaging
  * [Issue 1152255: ServiceWorker is shut down every 5 minutes for manifest V3 extension](https://bugs.chromium.org/p/chromium/issues/detail?id=1152255)
* Wasm
  * [Issue 1173354: wasm does not work in extensions manifest v3](https://bugs.chromium.org/p/chromium/issues/detail?id=1173354)

## Other discussions

* #11
* #16
* #25
* #44
* #51
* #39

**Comment by mingyaulee** [comment]
One additional use case for background service instead of service worker is the use of web assembly.
This is related to the issue I raised in #39.

**Comment by cuylerstuwe** [comment]
Not added here is extension-driven behavior from an external source.

This includes user-driven input coming from an outside _(related)_ desktop/cloud app via **WebSockets** or **Native Messaging Hosts**.

Their connections need to persist indefinitely in order to work effectively.

**Comment by cohosh** [comment]
Hey! I've just shared a use-case where we need access to RTCPeerConnection for service workers in order to upgrade to using v3 manifests here: https://github.com/w3c/webrtc-extensions/issues/77#issuecomment-912667886

**Comment by avi12** [comment]
Another use-case is using FFmpeg via [@ffmpeg/ffmpeg](https://github.com/ffmpegwasm/ffmpeg.wasm), which is using WebAssembly
To provide the best user experience using this module, you need to put it on the background page, which would make it accessible from all of the parts of the extension
Such a background page must also be persistent, as it takes a few seconds until FFmpeg gets initialized; if it isn't persistent, it might get terminated by the browser, thus requiring re-initializing FFmpeg for every user action that would result in FFmpeg processing

**Comment by dotproto** [comment]
> Such a background page must also be persistent, as it takes a few seconds until FFmpeg gets initialized; if it isn't persistent, it might get terminated by the browser, thus requiring re-initializing FFmpeg for every user action that would result in FFmpeg processing

@avi12, from the Chrome team's point of view terminating an unused worker is a feature, not a bug. Can you expand on why reinitializing FFmpeg is a problem? Adding latency to an uncommon operation is understandably undesirable, but so is keeping around an unused JS context.

Is there a specific workflow with an existing extension you can point to?

**Comment by dotproto** [comment]
@mingyaulee Added a crbug link related to allowing Wasm in MV3 service workers.

@cuylerstuwe, I just added a thread you started on chromium-extensions related to WebSockets and a crbug issue tracking native messaging.

@cohosh Added your link and the related crbug issue

**Comment by avi12** [comment]
> Is there a specific workflow with an existing extension you can point to?

@dotproto Right now I'm working on [this extension](/avi12/youtube-downloader), but because of the lack of a persistent background page and the other problems I mentioned, it would probably only work well as a manifest v2 extension

**Comment by avi12** [comment]
@dotproto 
> Adding latency to an uncommon operation is understandably undesirable, but so is keeping around an unused JS context.

That's true. However, in CLI apps like FFmpeg, which can only do one job at a time (which is particularly true for the WebAssembly version), if the job takes too much time (depending on the operation and the size of the files you work with), it might get terminated mid-way through, due to the limited 5-minute lifespan of an extension's Service Worker
> Is there a specific workflow with an existing extension you can point to?

 Right now, I'm working on [this extension](/avi12/youtube-downloader)

**Comment by avi12** [comment]
Another use-case is when needing to download files, which, if need to be initiated in the current Service Worker, will have 2 issues:
1. A Service Worker's lifespan is up to 5 minutes, If you need to download files that would take more than 5 minutes, it would be disrupted mid-way
2. If you find yourself needing to use the Fetch API (like in [my extension](/avi12/youtube-downloader)), making the download cancellable is important for a good user experience. The only way to cancel a Fetch process is by using [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)'s [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal), which so happens to use the DOM

**Comment by Jack-Works** [comment]
> A Service Worker's lifespan is up to 5 minutes, If you need to download files that would take more than 5 minutes, it would be disrupted mid-way

Try [Background fetch](https://developer.mozilla.org/en-US/docs/Web/API/Background_Fetch_API)?

**Comment by avi12** [comment]
Right, and where will it run? It needs to run either in a content script or in a Service Worker
If in a Service Worker, it imposes the problem I presented
If in a content script, it will stop running as soon as the user closes the tab

**Comment by avi12** [comment]
That's an interesting option
However, my extension that I linked above also needs to run FFmpeg on the background, so I don't think the Background Fetch API will be enough for my use-case

**Comment by cuylerstuwe** [comment]
In a world where RAM is plentiful and tends to be cheaper/easier to upgrade
than CPUs, it’s pretty odd that the current Chrome team would essentially
prioritize memory over CPU cycles.

This is especially weird since Chrome has historically performed better
than Firefox almost every synthetic and real-world benchmark, at the cost
of higher RAM use.

**Comment by tregagnon** [comment]
@dotproto I think [crbug 893175: Request for an API to specify Light and Dark Mode Icons](https://bugs.chromium.org/p/chromium/issues/detail?id=893175) should be added to the list

The Service Worker does not have `window.matchMedia()` which was used by Extensions to identify if the OS was in Dark/Light mode. We need the `action.setIcon()` API to be extended so Light/Dark icons can be set.

Defining those Light/Dark icons in the Manifest only is insufficient for extensions like [Dashlane](https://chrome.google.com/webstore/detail/dashlane-password-manager/fdjamakpfbbddfjaooikfcpapjohcfmg), which also uses the icon to indicates if the user is authenticated or not.

**Comment by ghostwords** [comment]
What about the general mismatch of webpages having user-defined lifetimes with Service Worker-based extensions having [arbitrarily limited](https://github.com/w3c/webextensions/issues/51#issuecomment-902613271) lifetimes ([five minutes in Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=1152255) at this time)?

The user may be relying on extension-provided functionality on some webpage when the extension's SW gets shut down. Even if the SW were to get restarted in response to content script messages, shutting down and restarting a SW is inherently a waste of resources. The additional delay caused by SW startup will break use cases that depend on speedy messaging between the content script and the background page (for example, an extension that dynamically modifies the right-click menu based on type of clicked element). Furthermore, after a SW restart, content scripts may have stale configuration data and won't work properly without additional message passing or a page reload.

If content scripts were to also get reloaded along with the SW, this would introduce clear usability issues, where the user is in the middle of some extension-facilitated workflow that suddenly stops working.

**Comment by tophf** [comment]
Another blocker: https://crbug.com/1219164 - **ability to create nested workers**.

BTW, the title should reflect that these use cases are currently impossible to implement in a service worker, please don't use the euphemistic language such as "are not well served".

**Comment by gildas-lormeau** [comment]
It's not possible to use `chrome.i18n.getMessage` to set the label of the action button or the entries in the context menu.

**Comment by capaj** [comment]
Safely storing confidential information for the duration when browser is running.

Chrome docs on storage api explicitly say that:
![image](https://user-images.githubusercontent.com/1305378/135106980-4f634014-8f67-4bc5-ad70-14524ff05052.png)

Yet they explicitly advise to use it to store data which is needed in the service worker:
https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/#state

I think @bitwarden relies on this when they store the master password inside backgroundPage so that the user doesn't have to put it in every time when they open the Popup.

**Comment by gorhill** [comment]
> Please use this issue to link to use use cases that are impossible to accomplish with or not well supported by service workers.

A use case not possible with a worker, ["Bug 1580254: Support CSS selector validation API from a service worker"](https://bugzilla.mozilla.org/show_bug.cgi?id=1580254):

> Removing access to a DOM would currently break uBlock Origin (“uBO”). uBO uses a DOM element to validate plain cosmetic filters[4] – which are essentially CSS selectors. Having access to the DOM to validate plain cosmetic filters allow uBO to detect and discard or further process invalid cosmetic filters.
> 
> [...]

**Comment by ghostwords** [comment]
>Requesting file:-scheme resources (aka file:// URLs)

The [Fetch API is also unable to retrieve `chrome://` resources](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/U6TpTj6C7ac/m/obB5j_MkBQAJ) apparently, something you can do now with `XMLHttpRequest`.

**Comment by tophf** [comment]
#### Dynamic import of scripts in a service worker is only partially implemented and prohibitively convoluted.

https://crbug.com/1198822

* No dynamic `import()`, also no `script` elements since there's no DOM.
* Only `importScripts` is supported.
  * it imports solely into the global scope, which means you would need a module management system i.e. use webpack (and similar systems) or find some service-worker-compatible library or write one yourself;
  * it's synchronous, which adds considerable internal overhead to stop the execution environment for the duration of reading the file. It's essentially the same as the ancient synchronous blocking XMLhttpRequest, deprecated and strongly discouraged for many years.

*  We have to duplicate `importScripts` explicitly inside the `install` event listener of the service worker for every script we will ever import later. The same will be necessary for the dynamic `import()` when it's implemented.

And there's also a foundational issue of an understandable reluctance in the service worker spec group to provide WebExtensions-specific improvements, further exacerbated by an understandable reluctance of the browser developers (Chromium extensions team, specifically) to provide non-standard exceptions in a service worker as it would add maintenance burden on them and introduce fragmentation of otherwise solid technology. Allowing the dynamic import of scripts without the need to pre-import them in the `install` event listener (all importable scripts are already local in an extension) is an example of such problem as well as the ability to control the lifetime of the service worker or create `blob:` object URLs, and many more problems introduced by the premature decision to switch to service workers in ManifestV3.

**Comment by avi12** [comment]
> from the Chrome team's point of view terminating an unused worker is a feature, not a bug. Can you expand on why reinitializing FFmpeg is a problem?

@dotproto Assuming the user is willing to wait a few seconds every time he wishes to use the FFmpeg functionality, the Service Worker will still get terminated after 5 minutes, making mv3 a _huge_ limitation

**Comment by Jack-Works** [comment]
> it imports solely into the global scope, which means you would need a module management system i.e. use webpack (and similar systems) or find some service-worker-compatible library or write one yourself;

For anyone interested, https://github.com/awesome-webextension/webpack-target-webextension/ this webpack extension support dynamic import in MV3 (early load chunks but not executing them).

**Comment by jonathanmayer** [comment]
Another use case where service workers are problematic: **maintaining time-sensitive state about browser windows, tabs, and pages**. My research group at Princeton has a number of academic studies that are implemented as browser extensions and that involve precisely understanding user interaction with the browser and web content. Even minor performance hiccups or event loop requeuing can introduce measurement error.

In the background page model (Manifest v2), we maintain state in variables and update state based on myriad API events and content script messages, using timestamps when possible. In the service worker model (Manifest v3), we can't rely on variables because the browser might shut down the service worker. The least-bad options, so far, are to either a) convert all of our (many) state variable operations to storage calls, which would introduce measurement error risks by littering asynchronous and slow storage calls throughout our implementation, or b) use runtime events to store state when the service worker shuts down and load state when the service worker starts again, also introducing measurement error risks and possibly failing because there's no guarantee asynchronous storage calls will succeed on shut down.

The option of a persistent extension worker, as many have proposed, would address this use case.

**Comment by bitcrshr** [comment]
Good day, everyone!

(Copied from Chromium Extensions forum by request)

I have taken on a new role in my company that involves me maintaining our browser extension which is based on MV2. I have a few concerns about the migration to MV3, and I was hoping to get some guidance.

For some background, we sell a bluetooth LE device that integrates with our password management software. Basically, as long as the device is in range, you can have access to your passwords.

The implication of this is that our desktop app needs to tell our Chrome extension "hey, the device is still in range!" regularly through a WebSocket. In fact, the extension's only way of getting sensitive data is through that WebSocket. I understand that WebSockets are still allowed in MV3, but I am concerned that the ephemeral nature of service workers will make this process impractical if not infeasible.

Sure, it's not a problem if the user doesn't need their credentials for a while, but what Chrome API could be called to 'wake up' the service worker after it dies? I've seen a good bit of workarounds in this forum talking about doing some arbitrary Chrome API call every 4.9 minutes to keep it alive or keeping a tab open specifically for the extension to run, but in my opinion these are hacky and counterintuitive.

Even if there was an MV3-compliant way to go about this, the user would no longer be able to see battery % or do basically anything at all in the extension if the service worker is not active.

I'll admit that I may not be getting the full picture here, but it seems that disallowing persistent background scripts in MV3 more or less breaks our entire Chrome platform. From what I have read around here, there are plenty of other people that have similar concerns and I hope that we will be able to find a solution that works for us all.

Also, I would like to politely note that the documentation for MV3 can be a little lacking--it's unclear how Chrome handles the lifecycle of Service Workers (i.e., conditions and timelines for becoming active/inactive). There's also not a lot of great guidance for maintaining state, other than the "in-memory storage" thing that doesn't seem to be a good replacement due to storage limits. All in all, I felt that a few questions that MV3 raised for me were far better answered in threads here than they were in the official docs.

I look forward to hearing y'alls thoughts on this!

**Comment by tylercrews** [comment]
I have an extension that I wanted to use as a kind of voluntary parental advisory software. 
The idea was for users to input problem websites, specify an allowed amount of time and a blocked amount of time, and create listeners for specified websites that would trigger onCompleted visits to that website to activate a declarativeNetRequest blocking them.

Although the net request blocking worked quite well, I needed to create listeners in response to users specifying websites, and unfortunately at the end of the service worker lifecycle the listeners are always deleted.

Additionally I wanted to use a badge to count down the amount of time remaining before the allowed and blocked periods end, using a setTimeout or setInterval.
Neither of these methods worked either, because the service worker is forced to die after 5 minutes, and so the badge eventually stops ticking. 

There was no satisfying workaround for either of these issues so I've since abandoned the project.

**Comment by birtles** [comment]
> Another blocker: https://crbug.com/1219164 - **ability to create nested workers**.

Building on this, I have an extension that, on install, downloads the necessary data files and stores them in Indexed DB using a worker so as not to block queries that occur while the update is still in progress. The process takes about 10 minutes (particularly on Chromium browsers where IDB performance is particularly poor) meaning that in addition to "nested" workers, a longer timeout is required.

[`absurd-sql`](https://github.com/jlongster/absurd-sql) appears to offer improvements to IndexedDB performance but it can _only_ be run in a separate Worker.

**Comment by klapauciusisgreat** [comment]
In general, as I understand it, anything depending on sign-in-with-Google/facebook, 2FA or any other auth method that requires a browser window to sign in will no longer work in Chrome extensions because auth would need to run in the service worker and need to have access to a window to open a popup. At least, I have seen no-one from the crx or firebase team being able to articulate a way forward.

If there is a way, it would be great to explain how to go about it.

**Comment by eadan-y** [comment]
Let's say I want my extension to send a message to the server every time a user switch tabs. 
I only want to log tab switches according to a remote configuration, which is fetched when the extension is loaded.
If the behavior is disabled, I don't want to register to the "tabs.onActivated" event at all (for performance reasons). 

With the service-worker approach, I will always "miss" the first tab change event, since every time the service worker will start - I will have to fetch the configuration first (either from cloud or the storage) in an async way, delaying the event registration. 

With the old background page (or a persistent service worker) - I could load the configuration just once, never missing any future event. 

Another thing - once I'm subscribed to "tabs.onActivated" event, there is no way I can "unsubscribe" it, so it will no longer start my service worker.

**Comment by aeharding** [comment]
Hi,

I was told to post there here:

https://groups.google.com/a/chromium.org/g/chromium-extensions/c/Qpr-Wf8BcsI

Quote below:

> Frankly, I'm at a loss of words. I understand the benefits of service workers for websites. They make sense.
>
> But there are so many web extension use cases that need persistent background pages.
>
> For example, I have an extension that acts like a password manager on steroids: It records the tab while the "session" is open on a tab, passes those recorded chunks to the server on an interval, and communicates with a websocket to the server to allow the session to be closed remotely, etc.
>
> It needs a persistent background page when these "sessions" are active.
>
> It's a corporate security product, and needs to be authoritarian over tabs it's been enabled on.
>
> There's simply no way to replicate its behavior with manifest v3. There's no way to store HTMLAudioElement or MediaStream in the background page for cleanup when the session ends. There's no way to have a background websocket. There's no way to automatically clear in-memory data when the browser is closed. There's no way to implement webRequestBlocking + onAuthRequired. There's no way to run an interval < every minute. There's no way to store Blob video chunks from the tabCapture MediaStream to queue to upload to the server. There's so much more that is just not there.
>
> I understand that most web extensions don't need this functionality. But some genuinely do. And I keep looking things up that my extension *needs*, and all I see are unanswered stackoverflow questions, empty forum threads, open issues.
>
> I don't know what to do. I'm sorry for this rant, but I feel like there's major chasms of functionality that are lost in v3. And it's not just one thing, it's many completely absent core features to my extension's functionality.

Thanks.

**Comment by wolfspelz** [comment]
weblin.io is a messenger that lets people meet with avatars on web pages. Uses XMPP on websockets. A limited background/webworker lifetime would let users leave/enter all the time. The background page is the right abstraction: a persistent context that manages and coordinates states of multiple tabs AND holds the shared XMPP connection. A connection for each tab would generate much more load on client and server, would make the experience sluggish, would result in lots of TCP reconnects, and would need awkward workarounds. Might well kill the project. See issue #133

Now I install the extension and then it works seamlessly. That's what users want. 

They do not want an extra tab that is automatically. You write: an additional tab is "an imperfect fit. Note quite as utilitarian, but it gets the job done". I am afraid this means: it is a worse user experience. If closed accidentally then either weblin.io breaks or auto-opens another tab which sounds annoying and error prone. Closing a tab sounds like a nice way to disable it but disabling is currently done by a single click on the extension icon. Cannot be easier. The tab is a workaround that makes things worse. It is only necessary because the good solution goes away. 

The native companion app with native messaging looks good on paper if you ignore the fact that people must download a native app in addition for things to work. I had a native app before and switched to the extension because it is easier to convince people and safer for them. That's what the browser sandbox is good for. I don't understand why you would take that away making things more difficult for users, for the business, and by requiring native code less safe from the user's point of view. The native app / native messaging approach is only useful if the native app is the main and the extension is the addon. 

The problems with handling the messaging connection on the page (iframe or directly) are obvious: a new connection on each navigate, takes more time to set up, needs more resources. That is generally considered so bad that HTTP 1.1 was developed to add keep-alive. Also (not as obvious) prohibits an intelligent mechanism, that skips leave/enter room during navigations, so that the avatar does not flash all the time from the point of view of other visitors. That makes it much worse for others. 

If I only knew the rationale for this change we might be able to find a solution that meets the rationale and still does not make our app worse.

**Comment by avi12** [comment]
I encourage you all to state your issues in the official Chrome Dev Summit Discord server,  https://discord.gg/UhgDwtmw

**Comment by wolfspelz** [comment]
Any idea where exactly. I posted something in #extensions 
Can I do more?

**Comment by avi12** [comment]
@wolfspelz You have #general

**Comment by wolfspelz** [comment]
Who drives this standardization? and why?

**Comment by wolfspelz** [comment]
With respect to [weblin.io](https://www.weblin.io) and Issue #133

I understand the desire to run extensions in resource constrained environments, especially if that means extensions will be available on chrome/android. Currently I am recommending my users to install another chromium-based browser on mobile. But you are probably aware that projects which need a persistent background context let the worker restart immediately. So, the worker is still running all the time despite the time limit. Only to restart it consumes more resources. Kind of defeats the purpose.

There is a difference between a compute job that runs for a long time in the background and a persistent context like a messenger connection. You are trying to kill permanent background computing by removing background persistence. But compute jobs that want to run permanently will just restart and applications needing a persistent context will need more resources to cope with the time limit. So, this change is not achieving its goal in one case and is actually counterproductive in the other.

**Comment by Anarchotect** [comment]
Say you use webRequest in a manifest v3, policy-installed, enterprise extension. Each request is either canceled, redirected, or allowed, but the decision has to be made based on the result of an API request, which you then cache for similar requests in the future.
```
chrome.webRequest.onBeforeRequest.addListener(
    (details) => mustBeSynchronous(details), { urls: [ '<all_urls>'] }, ["blocking"]
);

mustBeSynchronous(details) : chrome.webRequest.BlockingResponse {
    // Must use async fetch, cannot use sync XMLHttpRequest
    fetch('https://our.api.endpoint/param?uuid='+ourLookup)
        .then(response => response.json())
        .then(data => {
            //This returns a promise anyway, not a chrome.webRequest.BlockingResponse
            return makeDecisionsAbout(data);
        });
    // Returns as soon as fetch is sent, before promise resolves.
    //  tooLate cannot depend on makeDecisionsAbout()
    return tooLate;
}
```
Without a synchronous request, the page will load before a decision can be made. On subsequent requests a cached result can be used synchronously and locally to make the decision, but the first load of any page will bypass our logic.

I understand that asynchronous requests are better for performance, but in this instance we specifically need a synchronous request for exactly the reason it's shunned; we need to hold things up. I understand discouraging the use of synchronous requests, but banning the function entirely seems a bit Procrustean.

**Comment by markon1** [comment]
geolocation API is not working with service workers...

**Comment by avi12** [comment]
> geolocation API is not working with service workers...

You're welcome to report it on the official [Discord server](https://discord.gg/UhgDwtmw)

**Comment by dotproto** [comment]
> Who drives this standardization? and why?

@wolfspelz, the short answer is no one. There are currently no standards for browser extensions or groups pursuing standardization. This group is not an exception. While we are aiming "to identify common ground, bring implementations into closer alignment, and chart a course for future evolution" ([source](https://www.w3.org/community/webextensions/2021/06/04/forming-the-wecg/)), as stated in the [Community Group FAQ](https://www.w3.org/community/about/faq/#how-can-i-talk-about-w3c-endorsement-of-this-work), "Community Groups … do not create Web standards (though their outputs may advance to the standards track in a Working Group)."

**Comment by rektide** [comment]
>> *Who drives this standardization? and why?*

> *the short answer is no one. There are currently no standards for browser extensions or groups pursuing standardization.*

It's depressing in extreme that such a distant ghostly shadow of what we de-jure have today would be what we would standardize & force upon ourselves in the coming months or years.

The current status quo affords extensions a range of powers comparable to what a page can do. That seems like a respectable baseline. To say that an extension should have less powers than a page seems absurd & upside down. This issue, #72, is one of the critical points. Other losses, like #139, seem also in need of justification & explanation, but which have never been explained or justified either.

What is being created now is awful, is a shriveled husk of what extensions have historically been. This community group, if the current path is maintained, is actively working against the freedom & future of the web, & should dissolve until some better less anti-user-agency basis can be established. One group, Google, has presented an effort, and so far it has been an abject failure.

**Comment by ghostwords** [comment]
Observational webRequest has been broken in service workers in Chrome for over two years now:
https://bugs.chromium.org/p/chromium/issues/detail?id=1024211

Now if the bug did get fixed and webRequest events did restart the service worker, what does having to repeatedly restart over the course of typical browsing (for no good reason whatsoever other than [Chrome's hardcoded five minute limit](https://bugs.chromium.org/p/chromium/issues/detail?id=1152255)) do to Google's Manifest V3 improved performance claims? Who knows for sure because webRequest is still broken!

Just a reminder, we are apparently [about thirteen months away from when Manifest V2 stops working in Chrome](https://developer.chrome.com/docs/extensions/mv3/mv2-sunset/).

**Comment by aeharding** [comment]
god this is so stressful

**Comment by hunterharris1** [comment]
Would also like to reiterate that nested workers not being supported is a problem for my company's use case as well.

Added to the Chrome issue [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1219164). Quoted info below (w/ some extra details).

The company I work for has an extension that we use to augment the experience of our customers that use our analytics products. Basically we allow information overlaid in the form of cards for specific keywords that are tied to datasets they have chosen to be interested in.

To be able to effectively search the webpage we build an efficient data structure in a web worker from the background page (it is computationally expensive to build, but allows fast search time). We then serialize this and send that version to the content scripts to use. This significantly gives us a performance boost for the user’s experience and for the amount of time we have to borrow from an individual page’s resources.

Additionally, we also have a search feature, that leverages the use of lunr search indexes, and we use workers to build and search those indexes (again from the background page).

We will not be able to do either of these efficiently anymore with Manifest V3, and it looks like it will cause major performance issues for us.

Our users expect this to work on every page they are visiting, and a persistent extension really did make sense for our use case. 

By moving this to a service worker now, we now how to deal with large data serializations upon shutdown/startup, and w/o web workers have to do the computational heavy processing inline (or some major refactoring to do it in chunks, which will impact the perceived performance from the user's perspective).

Anyway, just posting here to help raise awareness of the limitations we're facing as it seems several others are having problems as well.

**Comment by cuylerstuwe** [comment]
> > Who drives this standardization? and why?
> 
> @wolfspelz, the short answer is no one. There are currently no standards for browser extensions or groups pursuing standardization. This group is not an exception. While we are aiming "to identify common ground, bring implementations into closer alignment, and chart a course for future evolution" ([source](https://www.w3.org/community/webextensions/2021/06/04/forming-the-wecg/)), as stated in the [Community Group FAQ](https://www.w3.org/community/about/faq/#how-can-i-talk-about-w3c-endorsement-of-this-work), "Community Groups … do not create Web standards (though their outputs may advance to the standards track in a Working Group)."

The more **accurate** answer is:

### Google

Chrome's extension platform is the "de-facto" standard which Firefox and Safari have sought to imitate to varying degrees. Chrome has most of the market share _(especially on desktop, where extensions are generally used)_, and implementors for other browsers understand that extension availability is a major draw to one browser or another. Extensions are the web's _"killer feature"_, and other browser implementors want to have access to Chrome's marketplace of extensions so that switching to those browsers doesn't cause so much suffering.

**A de-facto standard is still a standard, and by that token, developing the de-facto standard qualifies as "standardization".**

Google is mostly doing their own thing here, only really halfheartedly listening to any community feedback. The CG is mostly just Google just advertising what they're up to, while just about everyone who doesn't work at Google begs desperately for some sanity.

I think I speak for much of the developer community in hoping that this standardization effort fails, so that at least one major browser will actually remain worth using and developing for. It's totally pointless for all parties involved to standardize upon something inferior.

You remember when you were little and your mom asked you, "If your friend jumped off a bridge, would you?" Let Google jump off this bridge. Let Chrome become irrelevant through Google's own poor decisions. Let's watch and see what comes next. 🤷‍♂️

**Comment by wmbenedetto** [comment]
I'm looking at porting StayFocusd from Manifest v2 to v3, and it's pretty clear that moving from background pages to service workers is going to completely brick my extension. StayFocusd blocks time wasting websites once you have spent more than an allotted amount of time on them. To do this, it needs to run a setInterval() in a background page to check once a second how much time has elapsed. Once the time has expired, it blocks sites on the blocklist for the rest of the day. If I can't keep a persistent timer running, then my extension is useless. It was one of the first Chrome extensions, dating back to Jan 2010, and has over 625,000 users. The fact that it will be killed by this change is incredibly frustrating. There may be a hack workaround that I haven't found yet -- I just started looking into the changes -- but I shouldn't need to do that just to keep my extension alive.

**Comment by wolfspelz** [comment]
> I'm looking at porting StayFocusd from Manifest v2 to v3, ...

We collected many cases in which the new manifest effectively kills old extension, because it is either impossible or very time consuming to change the code. Even if it is technically possible most devs moved to other projects just maintaining existing extensions on a small budget or even spare time. That makes major rewrite impossible in many cases.  

So, if that move is to clear out un-maintained extensions, then a small change would do the trick without killing extensions that are still maintained on a small budget. 

If the move is to force extensions to use less background resources then it does not the trick, because re-written extensions will use the same or even more resources to cope without persistency.

I wish they would let existing extensions live with a small change (to prove that they are still maintained) and force only new extensions to very different model. My proposal:
- Make a V3 spec with small changes, for example require the use of V3 for all extensions but actually don't change the definition from V2. Then require V3 for updates. That proves that an extension is still maintained. 
- Make V4 with all the changes that are now in V3 and require new extensions to use V4. That makes new code use the new spec. Require V4 for new submissions.

**Comment by nopol10** [comment]
@wmbenedetto in StayFocusd's case is it possible to run the timer in each page's content script and have it read and update the local storage?

**Comment by tophf** [comment]
@nopol10, it's unreliable because content script timers aren't isolated by design so they can be cleared by the web page or by another extension or by the user in devtools console, https://crbug.com/860320.

**Comment by gorhill** [comment]
> page's content script and have it read and update the local storage?

Additionally, extensions should never ever write private extension data to a sites' local storage.

**Comment by nopol10** [comment]
I was referring to `chrome.storage.local` which I've seen https://github.com/hindmost/reduxed-chrome-storage use as a way to get Redux working in MV3. And it seems to be the new recommended way to pass state around the extension https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/#state though as others have pointed out, Google's own docs warn against storing confidential user information so use cases might be limited

**Comment by wmbenedetto** [comment]
@nopol10 That's the "hack workaround" I was referring to. It might be theoretically possible, but it will be waaaaay more complex and unreliable than the current implementation. With the background page, I have a single master timer that applies to all tabs and windows, so there's a single source of truth. With content scripts, I'd have a timer assigned to each tab, with no single source of truth. That means I'd have to keep track of all the different timers and make sure that only one is running at a time, creating and destroying timers every time a user navigates to a new page or opens a new tab. It will be a nightmare, and even then is still victim to all the issues that @tophf mentioned. 😞

**Comment by scholtzm** [comment]
It's January 20th, 2022 and according to Manifest v2 Support Timeline[1], we can no longer submit public Mv2 extensions. Yet, there are many use cases which are not supported by the sheer nature of ephemeral service workers.

For example (among other broken APIs), our extension uses WebSockets to receive live updates from our servers. How can we implement this functionality with current Mv3 standard?
The fact that extensions are being made less powerful than a standard web page is beyond my comprehension.

[1] https://developer.chrome.com/docs/extensions/mv3/mv2-sunset/

**Comment by capaj** [comment]
@microsoft @mozilla I would very much like to hear what your official stand point is towards manifest V3 vs manifest V2. Which one do you prefer? Which one you will support going forward?
Because V3 seems like something @google has cooked up to kill ad blockers and raise their ad revenue.

@MicrosoftEdge are you going to kill manifest V2 as well? Are you going to drive more people towards browsers like @brave or @kiwibrowser on android?

**Comment by markon1** [comment]
it's impossible to integrate google analytics with MV3, since it won't allow this in `manifest.json`:
`"content_security_policy": { "extension_pages": "script-src 'self' https://www.google-analytics.com https://www.googletagmanager.com https://js.refiner.io; object-src 'self'" }`

**Comment by kzar** [comment]
> I'm looking at porting StayFocusd from Manifest v2 to v3...

Could it work like this?

1. `webNavigation.onCommitted` listener in background serviceWorker notes the time a distracting website is opened in storage (if there is not already a time noted for that website).
2. `chrome.alarms` listener fires every one minute and checks if any distracting websites are due to be blocked.
3. When a website is due to be blocked, a dynamic `declarativeNetRequest` rule is added to block that website.
4. When the `chrome.alarms` listener notices a day has passed, all `declarativeNetRequest` rules and logged times for distracting websites are cleared.

**Comment by tophf** [comment]
Edit: self-hidden due to being off-topic.

@kzar, it probably could, at the cost of reducing the precision to 1 minute (not a problem here I guess) and [quite possibly] restarting the background script hundreds of times a day, thus wasting battery and other resources [astronomically] more than the equivalent MV2 extension with a persistent background script, so the author's extension is yet another fine example of a bad fit for a service worker as it is defined now. We have to admit that persistence is not bad per se for some cases where the lack thereof is much worse for overall performance and resource consumption, and start specifying how the service worker's lifetime should be dynamically prolonged based on heuristics, measurements, histograms, statistics to achieve a realistically better outcome. There's an [evidence](https://crrev.com/c/2510431) that Chromium team has at least one developer who seems to understand this.

**Comment by kzar** [comment]
> ...restarting the background script hundreds of times a day, thus wasting battery and other resources...

Yea, good point about it being inefficient to fire the alarm every minute. I think [the chrome.alarms API also lets you pick a specific time](https://developer.chrome.com/docs/extensions/reference/alarms/#property-AlarmCreateInfo-when) for the alarm to fire. So I guess it could work something like this:

1. `webNavigation.onCommitted` listener creates an alarm when a distracting website is navigated to (if there isn't an alarm or `declarativeNetRequest` rule already created for the site).
2. When the alarm fires, a dynamic `declarativeNetRequest` rule is added to block that website.
3. A periodic alarm also fires every X hours or once a day and clears any other alarms or dynamic `declarativeNetRequest` rules.

_(FWIW I'm not defending the switch from background page to ServiceWorker, just trying to help.)_

**Comment by alanhkarp** [comment]
I have a password manager that holds the user's master password in the memory of the background page in Manifest 2.  Of course, that approach won't work with Manifest 3.  

My solution was to create a tab that runs what was the background script.  Every time the service worker starts, it looks for this tab and creates it if it doesn't exist.  Any state I put in the script that runs in the tab persists as long as the tab is open.  

I was concerned about the effect on the user experience, but so far it appears to be minimal.  My biggest problem is figuring out what to show in the tab that makes its presence seem natural.

The other issue is security.  I understood access to the background page, but I'm less certain about the code running in the tab.  Is there a risk to the user's master password because it's held in the memory of a script running on a web page?  A search didn't turn up anything, which both comforts and worries me.

**Comment by tophf** [comment]
@alanhkarp, chrome.storage.session should help you, it's yet undocumented, but the API is the same as chrome.storage.local with an overall limit of 1MB for the entire storage.

**Comment by alanhkarp** [comment]
Thanks for the pointer, @tophf, but I don't need session storage.  I have a running process.  Once I set the variable, I can get the value as long as the tab is open.

**Comment by tophf** [comment]
@alanhkarp, I was commenting on the first sentence in your comment:

> I have a password manager that holds the user's master password in the memory of the background page in Manifest 2. Of course, that approach won't work with Manifest 3.

chrome.storage.session is an API created specifically for this use case.

**Comment by alanhkarp** [comment]
Oops, @topf, my bad.  My code will be far simpler if I can use session storage for the master password.

I had the wrong mental model of session storage.  I thought it was connected to the session when the popup or service worker were open and would be lost after they closed.  

I just ran some experiments, and that is not the case.  I found an article that says "session" refers to the time a web page is open.  For an extension, that appears to be while the browser is open and the extension is enabled.  

I have all kinds of questions.  Is session storage for extensions documented anywhere?

One problem with session storage is that it is not recommended for sensitive data, such as passwords.  Is that true for extensions?  I'm not using any third-party scripts, just the ones included in my extension.

**Comment by cuylerstuwe** [comment]
chrome.storage.session is not the same as window.sessionStorage.

The former was poorly-named and unfortunately that poor name seems to have
stuck now. It would have been better to call it e.g.,
chrome.storage.ephemeral, chrome.storage.volatile, etc.

**Comment by alanhkarp** [comment]
Thanks for the clarification, @cuylerstuwe.  No matter what you call it, you still have to understand its lifetime.  I haven't found anything about that for the session storage of an extension.  For what it's worth, I ran my experiments with chrome.storage.session, but I should probably also see how window.sessionStorage works.

**Comment by tophf** [comment]
chrome.storage.session is an API created specifically for storing sensitive data in-memory without writing it anywhere, which is why it's almost just as safe as a variable inside a persistent background script, assuming the web page didn't find an exploit to break through the world isolation barrier. There's some info in https://crbug.com/1185226. When chrome.storage.setAccessLevel is implemented we will be able to hide the values from being directly accessible in content scripts so that a malicious web page would have to also break through the inter-process barrier, which is incomparably harder if even possible.

**Comment by scholtzm** [comment]
One catch, storage.session API is asynchronous which makes it impractical as a storage for data that was previously stored in memory in persistent background page. You might find yourself a) rewriting large parts of your code or b) hacking around it by greedily fetching all of the data[1] before you reach critical part of your code. This also means you won't be able to use `importScripts` in your SW.

[1] As suggested officially here: https://docs.google.com/document/d/1N34xDVIeIM2A6egjoKUQqjqcuQ2s2W8_v_H8ciFWBu4/edit

**Comment by tophf** [comment]
> you won't be able to use importScripts in your SW.

You can use importScripts asynchronously but you need to call it additionally in the `install` event handler, [example](https://stackoverflow.com/questions/66406672/chrome-extension-mv3-modularize-service-worker-js-file). This is necessary for the web SW because all dependencies must be ensured to be locally cached. It makes no sense for extensions, which are already local, which is yet another reason why SW was a poor choice made by Chromium people who aren't JS developers.

**Comment by scholtzm** [comment]
> example

Doesn't the example load (and execute) `some-complex-script.js` twice? Once in `oninstall` and once in the async handler.

**Comment by tophf** [comment]
Yes, it does and the reason is explained in my comment above. Note that the `install` event is triggered just once when the extension is installed/updated.

**Comment by alanhkarp** [comment]
chrome.storage.session will work for my extension, but there are some things I only learned by experiment.  For example, it appears that the popup and service worker have separate session storage, but the popup and tabs the extension creates have the same storage.  I also tried window.sessionStorage, which works fine in the popup, but window is undefined in the service worker.

Is this stuff written up anywhere?

**Comment by tophf** [comment]
@alanhkarp, chrome.storage.session is the same throughout the entire extension, `sessionStorage` is an unrelated thing.

**Comment by zombie** [comment]
Please keep this issue as a summary/list of use cases.  If you need to discuss something in more detail, a separate issue is probably a better place.

**Comment by alanhkarp** [comment]
You are right, @tophf.  I had a bug in my experiment.

I agree, @zombie.

**Comment by alanhkarp** [comment]
Don't worry, @zombie, this post explains problems when using Manifest 3  for my extension :)

I have been playing around with both session storage and an additional tab to hold the user's master password.  They both work fine -- most of the time.  The tests I did are a pain, because I have to wait until the service worker goes idle.  Of course, I can't use the browser while I'm waiting because page loads keep the service worker active :(

The first case uses session storage with the bulk of the code running in the service worker.  Say that the user loads a page with a password field, but doesn't do anything with it for long enough for the service worker to go idle.  Clicking on the password field tells the content script to send a message to the service worker.  The message wakes up the service worker, but the response doesn't come back until the service worker goes idle again, after which I get a "message port closed" error.  I get the same behavior if I load an unrelated page to wake up the service worker.

The second case uses a tab created by the extension to hold the user's master password and to run the bulk of the code.  That works fine unless the user closes the tab while the service worker is idle.  Now the password manager appears to be hung until the service worker wakes up.  In my current implementation, the service worker doesn't receive any messages, so a page load is the only way to wake it up.

**Comment by hunterharris1** [comment]
@alanhkarp FWIW, to make your testing easier, this page exists where you can stop/start service workers - chrome://serviceworker-internals/

**Comment by alanhkarp** [comment]
Thanks, @hunterharris1.  I was running out of things to do while waiting for the service worker to go inactive.

**Comment by alanhkarp** [comment]
After some changes, the service worker reliably wakes up when it should.  I don't know why it wasn't happening before, but my extension now works with service workers.  Thanks for all your help, both related to the actual topic of this thread and not.

**Comment by discobeta** [comment]
> 

You can kill the service worker manually (for testing) by using chrome://serviceworker-internals/

**Comment by discobeta** [comment]
> 

FYI one can kill the service worker manually (for testing) by using chrome://serviceworker-internals/

**Comment by scholtzm** [comment]
Are there any plans to introduce clipboard access in service worker?

**Comment by nodeticswww** [comment]
One of the most important features of a persistent background page is to enable pre-fetching and pre-processing of lots of data that is then quickly accessible by the WebExtension UI (in-memory meta data lookup tables). So basically data **caching** is really important for performance.

One example of that is an RSS reader (Feedbro).

In Feedbro, the background page is also responsible for **playing audio clips**, **parsing HTML/XML pages** (requires an efficient XML/HTML parser provided by DOMParser), showing **desktop notifications**, calling **external services** (like IFTTT) and holding various data structures in memory. Background page also needs timed events and concurrent scanning of several different URLs (feed sources).

With MV3 it seems impossible to implement Feedbro in any reasonable way. Another victim will be our PageProbe extension which relies on much of the same MV2 features.

**Comment by wolfspelz** [comment]
> One of the most important features of a persistent background page is to enable pre-fetching and pre-processing of lots of data that is then quickly accessible by the WebExtension UI (in-memory meta data lookup tables). So basically data **caching** is really important for performance.

I totally agree. This change intends to lower resource usage in order to support weak (mobile) devices better. Ironically it turns out to make things worse. Caching background pages will have to reload their data using even more resources or apps abandon cashing as optimization strategy entirely, also harming performance. Kindof defeats the original purpose.

**Comment by aeharding** [comment]
Anyone know the latest status on MV3? Is Google still ignoring developers?

**Comment by leoshusar** [comment]
My use case is keeping the state of opened tabs, I am using events like `tabs.(onCreated|onUpdated|onRemoved)`.
I had to wrap all my event handlers in a preloader that retrieves data from storage, executes the handler and then saves the data again, like this:

```js
let initialized = false;

const initialize = async () => {
  if (!initialized) {
    const data = await chrome.storage.local.get(['tabInfo', 'controlsHidden']);
    Object.assign(tabInfo, data.tabInfo);
    Object.assign(controlsHidden, data.controlsHidden);

    initialized = true;
  }
}

const initializer = initialize();

const handleEvent = async (func, ...args) => {
  await initializer;

  try {
    await func(...args);
  } finally {
    await chrome.storage.local.set({ tabInfo, controlsHidden });
  }
}

chrome.tabs.onCreated.addListener((...args) => handleEvent(handleTabCreated, ...args));
chrome.tabs.onUpdated.addListener((...args) => handleEvent(handleTabUpdated, ...args));
chrome.tabs.onRemoved.addListener((...args) => handleEvent(handleTabRemoved, ...args));
```

If the reason for short lived workers was so it can save some CPU cycles, my workaround is the exact opposite, because I don't think calling `storage.get` on almost every event and `storage.set` on every event has the same performance as simply keeping the data in variables.

**Comment by bobthebuilder1997** [comment]
Service workers do not provide access to geolocation api. This is frustrating especially when using them in background scripts.

**Comment by leo60228** [comment]
> Anyone know the latest status on MV3? Is Google still ignoring developers?

@aeharding Limited Event Pages (#134) and Offscreen Documents (#170) are two proposals to help with the limitations of service workers. The former has implementations in Safari and Firefox. The latter is backed by Chrome but to my knowledge not yet implemented in any browser. The [2022-03-03 meeting notes](https://github.com/w3c/webextensions/blob/main/_minutes/2022-03-03-wecg.md#meeting-notes) have some relevant discussion too.

**Comment by iamnottheway** [comment]
Is there any update on persistent service workers? Or is there any API that lets you listen to when service workers stop?

**Comment by aeharding** [comment]
@leo60228 Offscreen documents don't support `chrome.tabCapture` api, so it doesn't work for me. I feel like mv3 completely ignores `chrome.tabCapture`.

**Comment by tdriley** [comment]
`setTimeout` and `setInterval` (though frowned on) have many unavoidable use cases and cannot be relied upon in an extension Service Worker because they're cleared when it suspends. The `chrome.alarms` API doesn't really offer an alternative given this restriction documented on the `create` method at https://developer.chrome.com/docs/extensions/reference/alarms/#method-create:

> Chrome limits alarms to at most once every 1 minute but may delay them an arbitrary amount more. That is, setting delayInMinutes or periodInMinutes to less than 1 will not be honored and will cause a warning. `when` can be set to less than 1 minute after "now" without warning but won't actually cause the alarm to fire for at least 1 minute. To help you debug your app or extension, when you've loaded it unpacked, there's no limit to how often the alarm can fire.

**Comment by bershanskiy** [comment]
> `setTimeout` and `setInterval` (though frowned on) have many unavoidable use cases and cannot be relied upon in an extension Service Worker because they're cleared when it suspends.

In my experience, extensions can combine `setTimeout`/`setinterval` with `alarms` to achieve almost any desired outcome. Specifically, use `setTimeout`/`setinterval` for short intervals less than 5 minutes (arbitrary lifetime limit at some point enforced by Chromium) and set regular `alarms` for events more than a minute apart. Some use cases for short `setTimeout`/`setinterval` do not even require persisting beyond service worker lifetime: in-memory cache with regular timed clenups, time limits for lengthy operations initiated by service worker itself, etc. Could you please describe your specific use case and perhaps I can describe how to do it with `setTimeout`/`setinterval` and `alarms`?

**Comment by bershanskiy** [comment]
Another use case not served by service workers is language change detection: `language` and `languages` live under `navigator` so are accessible from service workers, but oddly enough `languagechange` event is delivered to `window`. So extension backgrounds can detect the current language but can not react to changes (without regular pulling, of course).

**Comment by tophf** [comment]
> extensions can combine setTimeout/setinterval with alarms to achieve almost any desired outcome

@bershanskiy, this won't work in Chrome where 1) the SW lives for just 30 seconds since the moment it was started, 2) its lifetime is not prolonged by new `chrome` events, which is a nasty regression compared to MV2 event pages, 3) chrome.alarms is throttled to 1 wake-up per minute. The only way it can work is if we use a nativeMessaging host to keep the SW alive or constantly re-open ports from other tabs, which is not guaranteed to be open and requires broad host permissions.

**Comment by carlosjeurissen** [comment]
> Another use case not served by service workers is language change detection: `language` and `languages` live under `navigator` so are accessible from service workers, but oddly enough `languagechange` event is delivered to `window`. So extension backgrounds can detect the current language but can not react to changes (without regular pulling, of course).

In serviceWorkers you can listen for languagechange using:
```js
self.addEventListener('languagechange', console.log);
```

or simply:
```js
addEventListener('languagechange', console.log);
```

This has been tested in Mozilla Firefox, Google Chrome and Opera with success.

**Comment by DavidJCobb** [comment]
Service workers will lead to inconsiderate and wasteful bandwidth usage. Consider, for example, an extension that allows a user to manage or search through personal content or collections stored on a website (e.g. "favorites," "followers," "lists," etc.) in ways that a site does not normally offer -- perhaps a site with relatively minimal APIs, as opposed to the Internet's major corporate-funded platforms; in essence, a site that you have to scrape in order to extend.

The _polite_ way to implement something like this would be to cache data in-memory whenever possible (possibly clearing that cache after enough time spent idle or off-site) to minimize how many redundant requests have to go over the network during a single browsing session. However, you can't cache _anything_ using service workers. The 1MB storage in `chrome.storage.session` is _tiny_; it'd be pathetically small _even if_ we could store space-efficient binary data e.g. DataView, which, of course, we can't. The other storage areas are inappropriate for caching during a single browsing session, and may have privacy implications (e.g. for incognito browsing) as well. (There would also be logistical difficulties in taking what's supposed to be a session-specific cache and placing it in persistent storage. At minimum, you'd have to go to the trouble of wiping your jury-rigged cache `onStartup` and rejecting/deferring all queries until that's complete. With how patchy WebExtensions APIs have been in my experience, I have zero confidence that trying this wouldn't lead to some huge unexpected obstacle.)

That leaves the _impolite_ way: repeatedly bombard the website with requests for the same information you just asked for three minutes ago. Cache nothing; precompute nothing; optimize nothing; write rampantly wasteful code, because it's the only thing you're _allowed_ to write.

The performance cost for the network requests can be larger than one might expect. I've been privately developing and testing some WebExtensions and working on getting them release-ready, and some of the sites I've been making add-ons for have anti-botting/anti-DDoS measures. Slamming them with tons of rapid-fire redundant requests is a _great_ way to get rate-limited and receive a slew of 503s. If you need to scrape, say, a few dozen pages to gather up and preload metadata about a user's saved content, you might have to throttle those requests to ensure reliability. That's fine when it's information that you can query just once or twice, store in a background page, and then reuse throughout the rest of a browsing session (perhaps manually updating the in-memory copy without extra network requests if the user makes changes through the website). With service workers, however, you can't usefully cache anything, so that throttling ends up adding a _substantial_ delay to a WebExtension's reactivity, because now the user has to wait on it _every time the service worker wakes up_.

**Comment by radiolondra** [comment]
It seems to me (tested) that calling the function below when any service worker starts, **keeps the service worker up forever** (that is, it doesn't go to sleep anymore). 
**Question**: **Is this (another) Chrome bug or I found a way to keep a service worker alive forever?**

-------------------------------------------->>> **Code Description**
The function (_StayAlive_), inside the service worker, connects to a named port and tries to send a message through it to a nonexistent listener (so it will generate errors) when the service worker starts.

While doing this, **SW is active and running** (ok, it has something to do, that is, it has to send a message through a port, but _this is my doubt about a possible Chrome bug, because the connect() is started by the service worker itself, not by some external tab/app/extension/whatever..._ ok, lets go on with the code description).

Because noone is listening, it generates a (catched and logged) error (in "onDisconnect" port event) and terminates.

But after 25 secs it does the same iter from start, **keeping SW active forever**.

**Note**: To perform reliable tests remember to not open any DevTools page. Use chrome://serviceworker-internals instead and find the log (Scope) of your test extension ID.

This is the code:
```
// In ServiceWorker.js
// -------------------
// Forcing service worker to stay alive by sending a "ping" to a port where noone is listening
// Essentially it prevents SW to fall asleep after the first 30 secs of work.

const INTERNAL_STAYALIVE_PORT = "Whatever_Port_Name_You_Want"
var alivePort = null;
...
// Call the function at SW start
StayAlive();
...

async function StayAlive() {
    var lastCall = Date.now();
    var wakeup = setInterval( () => {
        
        const now = Date.now();
        const age = now - lastCall;
        
        console.log(`(DEBUG StayAlive) ----------------------- time elapsed: ${age}`)
        if (alivePort == null) {
            alivePort = chrome.runtime.connect({name:INTERNAL_STAYALIVE_PORT})

            alivePort.onDisconnect.addListener( (p) => {
				if (chrome.runtime.lastError){
					console.log(`(DEBUG StayAlive) Disconnected due to an error: ${chrome.runtime.lastError.message}`);
				} else {
					console.log(`(DEBUG StayAlive): port disconnected`);
				}

				alivePort = null;
			});
        }

        if (alivePort) {
                        
            alivePort.postMessage({content: "ping"});
            
            if (chrome.runtime.lastError) {                              
                console.log(`(DEBUG StayAlive): postMessage error: ${chrome.runtime.lastError.message}`)                
            } else {                               
                console.log(`(DEBUG StayAlive): "ping" sent through ${alivePort.name} port`)
            }
            
        }         
        //lastCall = Date.now();
               
    }, 25000);
}
```

**Comment by cacoos** [comment]
If you came here looking for recording video/audio solution for MV3, you can use this example: https://github.com/wireworks-app/chrome-screen-recording

**Comment by Pauan** [comment]
I have created an extension which helps the user to manage tabs more easily.

It displays the user's tabs in a vertical list, it allows for organizing tabs with labels, it allows for searching through tabs, it allows for bulk operations on many tabs, etc.

It's designed for users who have a lot of tabs (hundreds, or potentially even thousands of tabs).

Because it needs to process a lot of tabs, and it needs to process those tabs quickly, it keeps the tab data alive in RAM with a persistent background page.

When a new tab event happens (because a tab was closed, or opened, or moved, etc.) it can very quickly process the event, because all of the tab data is kept in RAM.

This is a very natural and efficient design, and it mimics what the browser itself does (the browser has its own C++ representation of the tabs which is always kept available in RAM).

However, this is incompatible with service workers, because service workers can never be persistent.

Re-processing potentially thousands of tabs on every tab event is unacceptable for performance, it will slow down the browser and lead to a bad user experience.

**Comment by Pauan** [comment]
I have created a [Chrome extension](https://github.com/Pauan/SaltyBetBot) which automatically bets on the [saltybet.com](http://saltybet.com/) website (automatic betting bots are allowed by the SaltyBet rules).

In order to make the bets, it needs to look at past historical data. There is a LOT of historical data (258 MB of data).

This data is loaded and processed inside of a persistent background page. This processing takes about 39 seconds.

This is acceptable because the data only needs to be loaded one time, and then it can be reused over and over again for each match.

However, this is incompatible with service workers, because service workers cannot be persistent.

Re-loading the data over and over again is completely 100% unacceptable for performance.

Any use case that requires loading or processing large amounts of data is incompatible with service workers.

For example, if somebody wanted to create a [folding@home](https://foldingathome.org/?lng=en) extension for Chrome, it would not work because it needs to store and process large amounts of data.

**Comment by Pauan** [comment]
Cryptocurrency wallets are incompatible with service workers.

Cryptocurrency requires a lot of expensive computations (to generate keys, download the blockchain, etc.). This computation can potentially take several seconds, or even minutes.

Because the computation is so expensive, cryptocurrency extensions must cache the computation so that it only needs to be done one time.

However, because service workers cannot be persistent, that means it cannot cache the computation.

Here is an example:

https://chrome.google.com/webstore/detail/leo-wallet/nebnhfamliijlghikdgcigoebonmoibm

Because of the lack of persistent background pages, that extension ***opens a new tab and then does all of its computation inside of that tab***.

So the user must keep that tab open at all times in order to use the extension.

This is obviously a completely unacceptable user experience.

Persistent background pages do not have that problem, because it can simply cache the computation inside of the (invisible) background page.

With persistent background pages, the extension could use a browser action popup for its UI.

But because service workers are not persistent, the extension cannot use a browser action popup, it must use a browser tab, which is an inferior user experience.

Any extension that requires in-memory caching of expensive computations is incompatible with service workers.

As time goes on, more extensions will start to abuse tabs as a substitute for persistent background pages, requiring the user to keep the tab open at all times in order for the extension to function.

**Comment by alanhkarp** [comment]
Can you use session storage?  That's what I'm using for a very small amount
of data.

I'm doing it with my fingers crossed because I haven't found a definitive
statement that what's in session storage never appears on disk.

--------------
Alan Karp

**Comment by Pauan** [comment]
> Can you use session storage?  That's what I'm using for a very small amount of data.

No, cryptocurrency wallets take security VERY seriously, it is unacceptable to have even the tiniest chance of private secrets leaked. Very large amounts of money are involved.

Anything private must be kept exclusively in memory, which means the extension must use in-memory caching with persistent background pages (or abuse tabs as a substitute for persistent background pages).

Even if sessionStorage was guaranteed to be in-memory only, it can only store a small amount of data, so it cannot support every use case.

Also, sometimes the cryptocurrency calculations take longer than 5 minutes, which means Chrome will kill the service worker, which obviously breaks everything. Caching cannot solve that problem.

So any extension that needs to run a very expensive computation (which takes longer than 5 minutes) is incompatible with service workers.

**Comment by tophf** [comment]
Note that session storage [leaks into the content script context](https://crbug.com/1342046) in Chrome, which can be hacked by a web page via a side channel attack as they share the same physical process.

**Comment by radiolondra** [comment]
You could give a look to this:
https://stackoverflow.com/questions/66618136/persistent-service-worker-in-chrome-extension/75082732#75082732
https://github.com/radiolondra/ServiceWorker-Highlander
https://github.com/radiolondra/ServiceWorker-Highlander-DNA
I use Highlander to make background worker persistent forever in my MV3 browser extensions. It works.
I already have talked about it here: https://github.com/w3c/webextensions/issues/72#issuecomment-1379374344
when I was starting to work on it.

**Comment by alanhkarp** [comment]
In the bug report, the content script has to set up a listener on
chrome.storage for the it to learn what's in session storage.  I don't do
that.  Does that mean I'm safe?

--------------
Alan Karp

**Comment by Pauan** [comment]
> @alanhkarp Does that mean I'm safe?

Maybe, maybe not. It's possible for malicious code to gain access to the content script (either through a browser bug, or XSS, or something like [Spectre](https://en.wikipedia.org/wiki/Spectre_(security_vulnerability))).

So even if *you* don't set up an event listener, if the malicious code is able to set up an event listener then it is vulnerable.

**Comment by alanhkarp** [comment]
Thanks.  I feel a little better.  That kind of vulnerability would be
pretty devastating to a lot more than my extension and would be pretty hard
to protect against anyway.  If you could change the code of my content
script, you could just ask my service worker for all its secrets, and it
would just send them to you.

--------------
Alan Karp

**Comment by erosman** [comment]
I came across a simple issue which has become complicated in the background service worker.

**'prefers-color-scheme' detection in Chrome background service worker**

```js
window.matchMedia('(prefers-color-scheme: dark)').matches
// Uncaught (in promise) ReferenceError: window is not defined
```

The complicated [chrome.offscreen](https://developer.chrome.com/docs/extensions/reference/api/offscreen) option  isn't optimal when `window.matchMedia` has to be performed repeatedly to set `action.setBadgeBackgroundColor`.

**Comment by hanguokai** [comment]
> 'prefers-color-scheme' detection in Chrome background service worker ……

This has been discussed several times in Chromium forum, like [this](https://groups.google.com/a/chromium.org/g/chromium-extensions/c/D5Jg2ukyvUc/m/m4kysya9AAAJ). Even in the event page, it also has a problem. That is why it requires a better solution from the browser #229 .

**Comment by visionarylab** [comment]
The service worker seem to want to minimize battery usage with lesser cpu cycle (service worker die in 5 min) and memory holding.  

Some browser vendor might argue, there is ton of other tool for "automation" on web, other than browser, because safety issue (Other UI tools)

1) For memory, with api similar to localstorage, it can be solve by assign limited memory amount to every extension. (Or just hidden the complexity by following options permission.

2) Since permission is the most successful stuff we had to protect everyone, from firefox to android era. Potentially there is a option list from 128MB to 1GB we willing to give to every extension we installed.

- Most user dont use more than 30 extension, it should be ok to micro manage by people. Permission is the only successful way for teaching user's security (modern generation)

**Comment by ParticleCore** [comment]
Scary to see that even after all this time, the original major issues have yet to be addressed. I'm adding my issue into the mix as well in hopes that it contributes for something https://github.com/w3c/webextensions/issues/518

**Comment by cedricalfonsi** [comment]
I would like to bring attention to a challenge we're encountering with service workers in Chrome, specifically their inability to fetch from servers presenting invalid certificates: `Error: (failed)net::ERR_CERT_AUTHORITY_INVALID`.

This issue is documented in the following Chromium tickets:

- [Chromium Issue 40882068](https://issues.chromium.org/issues/40882068)
- [Chromium Issue 40869181](https://issues.chromium.org/issues/40869181)

I want to verify with you if fetching from servers with invalid certificates via the extension service worker will be supported over the long term. Additionally, could employing the offscreen API be a viable temporary workaround for this problem? Any guidance on this approach would be appreciated.

**Comment by scholtzm** [comment]
3 years later and we still can't use dynamic/postponed imports:

https://issues.chromium.org/issues/40760920

**Comment by JWorthe** [comment]
Today I was trying to debug a performance issue in the service worker of our extension which has recently migrated to MV3.

In Chrome's Devtools, on the Performance tab, it tells me "Performance trace recording not supported for this type of target".

**Comment by tophf** [comment]
> Today I was trying to debug a performance issue in the service worker of our extension which has recently migrated to MV3.

There's a terrible but functioning workaround:
1. open any page/file from your extension in a tab so that its URL is chrome-extension://id/....
2. duplicate this tab
3. open devtools in both tabs
4. in tab A -> devtools -> Application -> Service worker -> click `stop`
5. in tab B -> devtools -> Performance -> start recording
6. in tab A -> devtools -> Application -> Service worker -> click `start`
7. in tab B -> devtools -> Performance -> stop recording, drag the flamechart to reveal the service worker row, expand it

It can be simplified if your service worker has a listener that can be triggered externally e.g. chrome.action.onClicked, chrome.commands.onCommand, chrome.contextMenus.onClicked, or any other listener triggered in response to something like navigation in a tab:

1. open any page/file from your extension in a tab so that its URL is chrome-extension://id/....
2. devtools -> Application -> Service worker -> click `stop`
3. devtools -> Performance -> start recording
4. perform an action for which there is a listener in the service worker
5. stop recording, drag the flamechart to reveal the service worker row, expand it

An alternative method for stopping/starting the service worker is `chrome://serviceworker-internals` page, which is less convenient, but it doesn't close when you reload the extension or call chrome.runtime.reload().

**Comment by JWorthe** [comment]
Thank you so much for the workaround, @tophf!

**Comment by Fiavaion** [comment]
A service worker can't access WebGPU at all, so any extension running a local model (e.g. via [@mlc-ai/web-llm](https://github.com/mlc-ai/web-llm)) has to host inference in a chrome.offscreen document. That works, but it doesn't escape the lifecycle problem — it relocates it. A single generation can run for tens of seconds to minutes, and the service-worker idle timer plus the offscreen-document lifecycle can tear the context down mid-generation. The caller then receives a truncated result with no error signal — the same "long-running compute killed part-way through" failure @avi12 described with FFmpeg-WASM above, but with silent data loss rather than a re-init cost.

The accessibility angle makes the silence the real problem: when the local model is doing text simplification or summarisation for a user who depends on it, a silently-truncated output isn't a latency annoyance — it's a correctness failure the user has no way to detect.

I've written up the reproduction, the offscreen-document keep-alive/chunking mitigation, and the accessibility dimension in #1014 — linking it here so it's reachable from the canonical catalogue.
